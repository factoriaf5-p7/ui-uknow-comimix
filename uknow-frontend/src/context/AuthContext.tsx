import { createContext, useState, ReactNode, Dispatch, SetStateAction } from 'react';
import LoginData from '../interfaces/login.interface';
import { useLoginUser } from '../services/useMutation-LoginUser';
import { User } from '../interfaces/user.interface';

export interface AuthContextType {
  isLoggedIn: boolean;
  loginData: LoginData;
  setLoginData: Dispatch<SetStateAction<LoginData>>;
  login: () => void;
  logout: () => void;
  user: User  ;
}

const initialAuthContext: AuthContextType = {
  isLoggedIn: false,
  loginData: {
    email: '',
    password: '',
  },
  setLoginData: () => {},
  login: () => {},
  logout: () => {},
  user: {
    _id: '',
    name: '',
    last_name: '',
    email: '',
    wallet_balance: 0,
    created_courses: [],
    chat_notifications_sent: [],
    chat_notifications_received: [],
    profile: '',
    bought_courses: [],
    __v: 0
  }
}
export const AuthContext = createContext<AuthContextType>(initialAuthContext);

interface AuthContextProviderProps {
  children: ReactNode;
}

const AuthContextProvider = ({ children }: AuthContextProviderProps) => {
  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const [loginData, setLoginData] = useState<LoginData>({
    email: '',
    password: '',
  });
  
  const [user, setUser] = useState<User>({
    _id: '',
    name: '',
    last_name: '',
    email: '',
    wallet_balance: 0,
    created_courses: [],
    chat_notifications_sent: [],
    chat_notifications_received: [],
    profile: '',
    bought_courses: [],
    __v: 0,
  });

  const loginMutation = useLoginUser();
  const login = async () => {
    try {
      const { data } = await loginMutation.mutateAsync(loginData);
      
      const token = data.token; 
      const userData = data.user;

      localStorage.setItem('token', token);
      localStorage.setItem('user', JSON.stringify(userData));
      setIsLoggedIn(true);
      setUser(userData);
    } catch (error) {
      console.error(error);
    }
  };

  const logout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('user'); 
    setLoginData({ email: '', password: '' });
    setIsLoggedIn(false);
  };

  return (
    <AuthContext.Provider value={{ isLoggedIn, loginData, setLoginData, login, logout, user }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthContextProvider;