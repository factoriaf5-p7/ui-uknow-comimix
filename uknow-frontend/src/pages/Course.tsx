import Navbar from '../components/navbar/Navbar';
import { useLocation } from 'react-router-dom';
import CourseDetails from '../components/course/CourseDetails';
import Footer from '../components/footer/Footer';
import CourseHeader from '../components/course/CourseHeader';
import CourseContent from '../components/course/CourseContent';
import CommentsSection from '../components/course/CourseCommentsSection';
import CourseNewComment from '../components/course/CourseNewComment';


const Course = () => {

  const location = useLocation()
  const course = location.state;

  return (
    <>
    <div style={{padding : ' 0 0 7% 0'}}>
      <Navbar />
      <CourseHeader course={course}/>
      <CourseDetails course={course}/>
      <CourseContent course={course}/>
      <CommentsSection courseId={course._id}/>
      <CourseNewComment course={course}/>
      <Footer/>
      </div>
    </>
  );
}

export default Course;