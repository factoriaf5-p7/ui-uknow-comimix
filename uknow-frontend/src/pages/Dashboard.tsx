
import DashboardTab from '../components/elements/DashboardTab'
import Footer from '../components/footer/Footer'
import Navbar from '../components/navbar/Navbar'

function Dashboard() {
  return (
    <>
      <Navbar/>
      <DashboardTab />
      <Footer/>
    </>
  )
}

export default Dashboard