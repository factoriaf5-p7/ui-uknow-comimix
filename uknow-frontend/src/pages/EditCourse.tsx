import React from 'react'
import EditCourseForm from '../components/forms/EditCourseForm'

function EditCourse() {
  return (
    <EditCourseForm />
  )
}

export default EditCourse