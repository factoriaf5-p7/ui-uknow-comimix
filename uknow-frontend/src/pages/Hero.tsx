import { HeroImage } from "../components/elements/HeroImage"

function Hero() {
  return (
    <>
      <HeroImage/>
    </>
  )
}

export default Hero