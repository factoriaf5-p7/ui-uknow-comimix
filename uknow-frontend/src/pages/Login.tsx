import Footer from "../components/footer/Footer";
import {LoginForm} from "../components/forms/LoginForm";

export default function Login() {
  return (
    <div>
      <LoginForm />
      <Footer/>
    </div>
  )
}
