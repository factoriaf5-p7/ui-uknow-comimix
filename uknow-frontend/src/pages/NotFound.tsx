import { Link } from "react-router-dom"
import Footer from "../components/footer/Footer"

function NotFound() {
  return (
    <>
      <div style={{ display: 'flex', justifyContent: 'center', alignItems: 'center', flexDirection: 'column' }}>
        <h1>404 - NOT FOUND</h1>
        <h2><Link to='/'>Go Home</Link></h2>
      </div>
      <Footer/>
    </>
    )
}

export default NotFound