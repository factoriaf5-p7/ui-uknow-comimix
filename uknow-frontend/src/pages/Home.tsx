import { useContext, useState } from 'react';
import CourseList from "../components/course/CourseList";
import SearchBar from '../components/elements/SearchBar';
import uknowImg from '../assets/uknow.png';
import { Container } from "@mui/system";
import { useAllCourses } from "../services/useQuery-AllCourses";
import { CourseData } from '../interfaces/course.interface';
import Navbar from '../components/navbar/Navbar';
import Footer from '../components/footer/Footer';
import { AuthContext } from '../context/AuthContext';

function Home() {
  const { isLoggedIn, user } = useContext(AuthContext); 
  const { isLoading, isError, courseList: allCourses } = useAllCourses();
  const [courses, setCourses] = useState<CourseData[]>([]);
  const [isSearching, setIsSearching] = useState(false);
 

  // Function to filter out courses that the logged-in user has already bought or created
  const filterUserCourses = (courses: CourseData[] | undefined) => {
    if (!isLoggedIn || !user || !courses) {
      return courses; // If no user is logged in, return all courses
    }

    // Filter out courses that the user has already bought or created
    const filteredCourses = courses.filter(course => {
      const isBought = user.bought_courses.some(
        (boughtCourse: { course_id: string }) => boughtCourse.course_id === course._id
      );
      const isCreated = user.created_courses.includes(course._id);
      return !isBought && !isCreated;
    });
  
    return filteredCourses;
  };

  // Handle the search results as before
  const handleAllCourses = (searchResults: CourseData[] | undefined = []) => {
    const filteredSearchResults = filterUserCourses(searchResults);
    if (filteredSearchResults) {
      setCourses(filteredSearchResults);
      setIsSearching(true);
    } else {
      console.error("Error: filteredSearchResults is undefined.");
    }
  };
  
  

  // Filter the courses based on the logged-in user data
  const filteredCourses = filterUserCourses(allCourses);


  return (
    <div>
      <Navbar />
      <img src={uknowImg} alt="Uknow image" style={{ width: '100%', height: 'auto' }} />

      <Container sx={{ display: 'flex', justifyContent: 'center', my: 2 }}>
        <SearchBar onSearch={handleAllCourses} />
      </Container>

      {isLoading ? (
        <div>Loading...</div>
      ) : isError ? (
        <div>Error fetching data</div>
      ) : (
        <CourseList courses={isSearching ? courses : filteredCourses} />
      )}
      <Footer />
    </div>
  );
}

export default Home;