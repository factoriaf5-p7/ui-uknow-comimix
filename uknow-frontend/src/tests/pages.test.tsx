import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import Course from "../pages/Course";
import { renderToString } from "react-dom/server";
import Hero from "../pages/Hero";
import Login from "../pages/Login";
import NotFound from "../pages/NotFound";
import Profile from "../pages/Profile";
import Recover from "../pages/Recover";
import Register from "../pages/Register";
import UnderConstruction from "../pages/UnderConstruction";
import Dashboard from "../pages/Dashboard";
import Home from "../pages/Home";
import { MemoryRouter } from "react-router-dom";
import { AuthContext, AuthContextType } from "../context/AuthContext";
import { User } from "../interfaces/user.interface";
import { CourseData } from "../interfaces/course.interface";
import EditCourse from "../pages/EditCourse";

const RouterWrapper = ({ children }: any) => {
  return <MemoryRouter>{children}</MemoryRouter>;
};
const mockAuthContextValue: AuthContextType = {
  isLoggedIn: true,
  loginData: {
    email: "",
    password: "",
  },
  setLoginData: () => {},
  login: () => {},
  logout: () => {},
  user: {
    _id: "1a",
    name: "john",
    last_name: "doe",
    email: "john@example.com",
    wallet_balance: 1000,
    created_courses: [],
    chat_notifications_sent: [],
    chat_notifications_received: [],
    profile: "yes",
    bought_courses: [],
    __v: 0,
  } as User,
};



  const mockCourse: CourseData = {
    _id: "mockCourseId",
    name: "string",
    price: 100,
    topic: "string",
    difficulty: "string",
    tags: [],
    bought: true,
    content: "string",
    createdAt: "string",
    updatedAt: "string",
    description: "string",
    image: "string",
    average: 3,
  };


const AuthContextMockProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  return (
    <AuthContext.Provider value={mockAuthContextValue}>
      {children}
    </AuthContext.Provider>
  );
};

describe("Checks if the pages render the pages correctly", () => {
  test("Snapshot test for Course page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
      <AuthContextMockProvider> 
          <RouterWrapper>
            <Course/>
          </RouterWrapper>
        </AuthContextMockProvider>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });
  test("Snapshot test for Dashboard page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <AuthContextMockProvider>
            <Dashboard />
          </AuthContextMockProvider>
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Hero page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Hero />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Home page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Home />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Login page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Login />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for NotFound page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <NotFound />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Profile page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Profile />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Recover page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Recover />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Register page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <Register />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });

  test("Snapshot test for Under Construction page", () => {
    const queryClient = new QueryClient();

    const html = renderToString(
      <QueryClientProvider client={queryClient}>
        <RouterWrapper>
          <UnderConstruction />
        </RouterWrapper>
      </QueryClientProvider>
    );

    expect(html).toMatchSnapshot();
  });
});
