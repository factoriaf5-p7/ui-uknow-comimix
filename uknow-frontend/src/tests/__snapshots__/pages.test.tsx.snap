// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checks if the pages render the pages correctly > Snapshot test for Under Construction page 1`] = `"<div style=\\"display:flex;justify-content:center;align-items:center;flex-direction:column\\"><h1>Page Under Construction</h1><p>This page is currently under construction. Check back later for more information.</p><h2><a href=\\"/home\\">Back Home</a></h2></div>"`;
